internal/fnruntime/container.go

```diff
@@ -158,20 +161,29 @@ func NewContainerEnvFromStringSlice(envStr []string) *runtimeutil.ContainerEnv {
 // exist.
 func (f *ContainerFn) prepareImage() error {
 	// check image existence
-	args := []string{"image", "ls", f.Image}
+	foundImageInLocalCache := false
+	args := []string{"image", "inspect", f.Image}
 	cmd := exec.Command(dockerBin, args...)
 	var output []byte
 	var err error
-	if output, err = cmd.CombinedOutput(); err != nil {
-		return &ContainerImageError{
-			Image:  f.Image,
-			Output: string(output),
-		}
-	}
-	if strings.Contains(string(output), strings.Split(f.Image, ":")[0]) {
+	if _, err = cmd.CombinedOutput(); err == nil {
 		// image exists locally
+		foundImageInLocalCache = true
+	}
+
+	// If env var ImagePullPolicyEnv is set to "ifnotpresent", we scan the local
+	// images first. If there is a match, we just return. This can be useful for
+	// local development to prevent the remote image to accidentally override
+	// the local image when they use the same name and tag.
+	if strings.ToLower(os.Getenv(ImagePullPolicyEnv)) == "ifnotpresent" && foundImageInLocalCache {
 		return nil
 	}
+
+	// If env var ImagePullPolicyEnv is not set, we will try to pull the image
+	// regardless if the tag has been seen in the local cache. Because we want
+	// to ensure we have the latest release for "moving tags" like v1 and v1.2.
+	// The performance cost is very minimal, since `docker pull` checks the SHA
+	// first and only pull the missing layer(s).
 	args = []string{"image", "pull", f.Image}
```

Comment:
Do you know about `docker run --pull`. You seem to be re-implementing that feature?

Reply:
We cannot let `run` to pull the images because it will print the pulling information to stderr and cannot be disabled.

Reply:
not sure I follow. We can control what to do with stdout and stderr of spawned processes.

Reply:
`docker run` will generate contents in stderr and we cannot distinguish it from the error messages written by function. So the result is user will see pulling progress in the stderr if function fails.

Reply:
If you must re-implement https://github.com/docker/cli/pull/1498/
do you want to at least use the same options?

Reply:
@mengqiy

Reply:
I will verify if this will pollute the output. If it does, I guess we still need to do this.

Also the current naming follows the k8s naming: Always and IfNotPresent.

Reply:
@Shell32-Natsu is right. If we use `docker run --pull=aways`, the stderr is kind of pullated by the message of pulling image. For example:
```
$ kpt fn render
Package "nginx": 
[RUNNING] "gcr.io/kpt-fn/set-labels:unstable"
[PASS] "gcr.io/kpt-fn/set-labels:unstable"
[RUNNING] "gcr.io/kpt-fn/kubeval:unstable"
[FAIL] "gcr.io/kpt-fn/kubeval:unstable"
  Results:
    [ERROR] Invalid type. Expected: [integer,null], given: string in object "apps/v1/Deployment/" in file "deployment.yaml" in field "spec.replicas"
  Stderr:
    "unstable: Pulling from kpt-fn/kubeval"
    "Digest: sha256:2f55da69f64ba0c12707fc6a57e54d1e31d7e900dd26de869397d079f3b22ab3"
    "Status: Image is up to date for gcr.io/kpt-fn/kubeval:unstable"
    ""
  Exit code: 1
```